Garden Tweaks & Features

WannaBeIan opened the pull request:
- [x] Added an option to view required crops in the visitor helper as stacks + remainder.

- [x] Visitor Helper now works outside of the Garden by default but includes a new setting to lock it to the Garden.

- [x] Fixed the regex for Farming XP so it works correctly and now displays your XP/h, even at max farming level.

- [x] Added decimals for more precise XP/h and Blocks/s calculations.

- [x] Visitors requesting the same crops are grouped together, displaying a sum of the required crops.

![New Settings](https://i.imgur.com/BAs39JB.png)

![Stacks + Grouped Visitors](https://i.imgur.com/cZsyU3N.png)

![Fixed XP/hr Counter + Fancy Dots](https://i.imgur.com/l6NDg6j.png)

I am aware the code is pretty messy its been some time since I've sat down and type something up

Let me know if anything needs to be fixed, and I'll take care of it!

kevinthegreat1 commented on src/main/java/de/hysky/skyblocker/skyblock/garden/VisitorHelper.java:
```diff
@@ -50,55 +47,134 @@ public class VisitorHelper {
-        }
-    }
+	// The location of copy amount, and item text seem to be overlapping so clicking on part of the itemText will run copy amount.
+	public static void onMouseClicked(double mouseX, double mouseY, int mouseButton, TextRenderer textRenderer) {
```
Definitely not a fan of the logic as is. I have a couple of suggestions:
1. You could use `SortedMap<String, Object2IntMap<Pair<String, String>>>` for `itemMap` instead, which won't keep one visitor's items together, but they usually only request one item anyways.
2. You could replace the mouse click logic with `ClickEvent` and `Screen#handleTextClick`. See how `ChatScreen#mouseClicked` does it. This way, you only have the rendering logic, and not the mouse clicked logic.

kevinthegreat1 commented on src/main/resources/assets/skyblocker/lang/en_us.json:
```diff
@@ -241,6 +241,12 @@
   "skyblocker.config.farming.garden.lockMouseTool": "Lock camera when holding a farming tool",
   "skyblocker.config.farming.garden.visitorHelper": "Visitor helper",
+  "skyblocker.config.farming.garden.visitorHelper.@Tooltip": "Makes it easier to manage visitors by grouping similar requests, copying item amounts, and quickly opening the Bazaar to buy items.",
+  "skyblocker.config.farming.garden.visitorHelperGardenOnly": "Visitor helper (Garden Only)",
```
I'd change this name to simply "Garden Only".

kevinthegreat1 commented:
After a discussion, we think the visitor helper options would be better in a group. If you can also do that it’d be great.

WannaBeIan commented on src/main/java/de/hysky/skyblocker/skyblock/garden/VisitorHelper.java:
```diff
@@ -50,55 +47,134 @@ public class VisitorHelper {
+		ScreenEvents.BEFORE_INIT.register((client, screen, scaledWidth, scaledHeight) -> {
+
+			String title = screen.getTitle().getString();
+			boolean isGardenLocation = Utils.getLocationRaw().equals("garden");
```
I've actually added a helper method into utils for this, but added that way after I made changes here. I'm getting close to pushing a final update for the VisitorHelper that also makes sure that the item map is linked based on the profileId so it isn't showing trades from other profiles.

I plan on eventually "revamping" it to make everything less of an eye sore but for now it'll work fine.

WannaBeIan commented:
Sorry for the messy commits. this should be everything requested change wise.
- VisitorHelper is now it's own group and work's/looks really good.
- Added a "Vacuum Solver" which just makes a guess based on the angry_witch partcles and places a waypoint at the location
- Added "Pest Highlighter" which applies a glow effect on pests.
- Made a little change in MayorUtils causing an error when no minister is found.
- Made the changes requested in the Farming hud, and when you're at max farming rather than it showing xp/0 I have farmingXpPercentProgress clamped between 0/100
- Noticed while going through the config that the description of suppliesAndFuelWaypointType was using an old translation so I updated that as well

Emirlol commented on src/main/java/de/hysky/skyblocker/utils/mayor/MayorUtils.java:
```diff
@@ -92,10 +92,17 @@ private static void tickMayorCache() {
-					minister = new Minister(ministerObject.get("key").getAsString(),
-							ministerObject.get("name").getAsString(),
-							new Perk(ministerPerk.get("name").getAsString(), ministerPerk.get("description").getAsString()));
+					if (ministerObject != null) { // Check if ministerObject is not null stops NPE caused by Derpy
+						JsonObject ministerPerk = ministerObject.getAsJsonObject("perk");
+						minister = new Minister(
+								ministerObject.get("key").getAsString(),
+								ministerObject.get("name").getAsString(),
+								new Perk(ministerPerk.get("name").getAsString(), ministerPerk.get("description").getAsString())
+						);
+					} else {
+						LOGGER.info("[Skyblocker] No minister data found for the current mayor.");
+						minister = Minister.EMPTY;
+					}
```
This seems a little redundant, as any exception thrown would land in the catch block which results in the same thing.

This does clarify it a bit more, so it could be kept. But the catch block's warn message must also be changed as it can no longer be caused by mayorless ministers and could cause confusion when something goes wrong.

Emirlol commented on src/main/java/de/hysky/skyblocker/skyblock/garden/FarmingHud.java:
```diff
@@ -35,119 +35,147 @@
-    }
+	private static final Logger LOGGER = LoggerFactory.getLogger(FarmingHud.class);
+	public static final NumberFormat NUMBER_FORMAT = NumberFormat.getInstance(Locale.US);
+	private static final Pattern FARMING_XP = Pattern.compile("§3\\+(?<xp>\\d+(?:\\.\\d+)?) Farming \\((?<percent>[\\d,]+(?:\\.\\d+)?%|[\\d,]+/[\\d,]+)\\)");
```
I'd also recommend `Formatting.strip`ping the message and removing the color code from the regex.

WannaBeIan commented:
The VacuumSolver could be improved by checking the scoreboard location line for the pest icon before rendering the waypoint, instead of rendering it regardless of pests. This can be addressed in a future PR.

kevinthegreat1 reviewed (approved):
LGTM. I've made some clean up and fixes if you want to take a look.

AzureAaron commented on src/main/java/de/hysky/skyblocker/config/configs/FarmingConfig.java:
```diff
@@ -30,6 +36,17 @@ public static class Garden {
         public boolean closeScreenOnPlotClick = false;
     }
 
+	public static class visitorHelper {
```
`VisitorHelper`

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/garden/VacuumSolver.java:
```diff
@@ -0,0 +1,120 @@
+		WorldRenderEvents.AFTER_TRANSLUCENT.register(VacuumSolver::renderTrail);
+	}
+
+	private static void reset() {
```
Shouldn't this reset the destination too?

kevinthegreat1 replied:
I tried to use this solver but couldn't figure it out so @WannaBeIan.

WannaBeIan replied:
Like I mentioned the code for the Vacuum Solver is really bad I mean it "kinda" worked it would get the particles and find where they end but it wouldn't estimate where the pest was I looked at some code from skyhanni as reference and that's where I got with it.

LifeIsAParadox replied:
I tested the Vacuum "Helper", but to be honest, it raised more questions than it actually helped. Maybe it just didn’t work as intended for me, but I think the logic could either be improved or perhaps the helper isn’t needed (at least for now).

Also, the red cube should probably disappear after some time or when switching worlds, otherwise, it just stays there permanently, which might not be ideal.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/garden/visitor/Visitor.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public record Visitor(Text name, ItemStack head, Map<Text, Integer> requiredItems) {
```
Should use an `Object2IntMap`

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/garden/visitor/VisitorHelper.java:
```diff
@@ -0,0 +1,259 @@
+		context.getMatrices().push();
+		context.getMatrices().translate(0, 0, 500);
+
+		for (Map.Entry<Text, Integer> entry : groupedItems.entrySet()) {
```
You need to use the type specific iterator

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/garden/visitor/VisitorHelper.java:
```diff
@@ -0,0 +1,259 @@
+		int index = 0;
+		int yOffsetAdjustment = -5;
+
+		for (Map.Entry<Text, Integer> entry : groupedItems.entrySet()) {
```
Type specific iterator

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/garden/VacuumSolver.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	private static void renderTrail(WorldRenderContext context) {
+		if (fixedDestination == null) {
```
Maybe also check if the person is in the garden and stuff?

kevinthegreat1 commented:
The vacuum solver needs a bit of testing and the potential bugs Aaron pointed out. I wasn't able to figure out how to use it.

LifeIsAParadox commented on src/main/java/de/hysky/skyblocker/skyblock/garden/visitor/VisitorHelper.java:
```diff
@@ -0,0 +1,259 @@
+				.map(String::trim)
+				.takeWhile(lore -> !lore.contains("Rewards"))
+				.filter(lore -> lore.contains(" x"))
+				.map(lore -> lore.split(" x"))
+				.forEach(parts -> visitor.addRequiredItem(Text.literal(parts[0].trim()), Integer.parseInt(parts[1].trim())));
```
`Caused by: java.lang.NumberFormatException: For input string: "1,024"`
crash lol
Full log (https://mclo.gs/BQSI0sc)

Maybe 
```diff
- Integer.parseInt(parts[1].trim())
+ Integer.parseInt(parts[1].trim().replace(",", ""))
```
 could help

kevinthegreat1 replied:
Fixed in 57f07c1.

LifeIsAParadox reviewed (changes requested):
First of all, thanks for the new features you added! I really appreciate the effort you put into improving the project.

I just have one small nitpick regarding the Vacuum Solver, and I also encountered a crash during testing.

Other than that, great work!

kevinthegreat1 reviewed (approved):
I've temporarily removed the vacuum solver, and you can open another pr with it once you fix the bugs. We'll merge this first.

kevinthegreat1 commented:
Please let me know if you need help with getting the vacuum solver back to work on it further.
